### PR TITLE
Fixes #2595 - Add more Rubyipmi methods to BMC API

### PR DESCRIPTION
**Pull request by Deltik**
Replaces: theforeman/smart-proxy#505
Replaces: 051b285aaa9b595ea9513ff5fa197eb830b942c1

This commit implements the same features as
051b285aaa9b595ea9513ff5fa197eb830b942c1, with some differences:

 - Added unit tests for all new features
 - Added interfaces to `modules/bmc/base.rb`
 - Cleared up some ambiguous method names
 - Changed some unit test stubs to expects
 - Fixed bugs in the original implementation

New API methods:

 - GET /bmc/:host/fru
   Shows available actions: "list", "serial", "manufacturer", "model",
   and "asset_tag"

 - GET /bmc/:host/bmc
   Shows available actions: "info", "guid", and "version"

 - GET /bmc/:host/sensors
   Shows available actions: "list", "count", "names", "fanlist",
   "templist", and "get"

 - GET /bmc/:host/lan/snmp
   Shows BMC host's SNMP community string

 - GET /bmc/:host/lan/vlanid
   Shows BMC host's VLAN ID

 - GET /bmc/:host/lan/ipsrc
   Shows whether BMC host's LAN was configured "static" or "dhcp"

 - GET /bmc/:host/lan/print
   Shows BMC hosts's LAN configuration

 - PUT /bmc/:host/chassis/power/reset
   Performs power reset on chassis

 - GET /bmc/:host/fru/list
   Shows all of host's FRU information. Includes workaround for
   undocumented IBM/Lenovo IPMI bug.

 - GET /bmc/:host/fru/serial
   Shows host's serial number

 - GET /bmc/:host/fru/manufacturer
   Shows host's manufacturer

 - GET /bmc/:host/fru/model
   Shows host's product model

 - GET /bmc/:host/fru/asset_tag
   Shows host's asset tag string, if Rubyipmi has this method. It's in
   the spec, but doesn't seem to be implemented today. See spec [here](https://github.com/logicminds/rubyipmi/blob/master/spec/unit/ipmitool/fru_spec.rb).

 - GET /bmc/:host/bmc/info
   Shows information about the BMC

 - GET /bmc/:host/bmc/guid
   Shows system's GUID

 - GET /bmc/:host/version
   Shows BMC's version

 - PUT /bmc/:host/bmc/reset?type=warm
   Performs warm reset on BMC

 - PUT /bmc/:host/bmc/reset?type=cold
   Performs cold reset on BMC

 - GET /bmc/:host/sensors/list
   Shows all of host's sensors

 - GET /bmc/:host/sensors/count
   Shows number of sensors on host

 - GET /bmc/:host/sensors/names
   Shows list of sensor names on host

 - GET /bmc/:host/sensors/fanlist
   Shows all of host's fan sensors

 - GET /bmc/:host/sensors/templist
   Shows all of host's temperature sensors

 - GET /bmc/:host/sensors/get
   Shows available option "sensor" for this action

 - GET /bmc/:host/sensors/get/:sensor
   Shows information about sensor :sensor on host

### CODE REVIEW

**Comment by dmitri-d** [comment]
[test]

File: modules/bmc/ipmi.rb [hunk 1]
```diff
@@ -148,6 +153,124 @@ def netmask
...
+      # BMC reset
+      def reset(type='cold')
+        host.bmc.reset(type)
+      end
+
+      # print all FRU information
+      def frulist
+        frulist = host.fru.list
+        # Detect quirk on ipmitool
+        raise 'Quirk' if frulist.empty?
```
Comment (dmitri-d) [inline modules/bmc/ipmi.rb]: I find this hard to read/follow. Can 'quirk' happen with ipmitool only? If so, why not return the error message on line 218 right away?

File: modules/bmc/ipmi.rb [hunk 2]
```diff
@@ -148,6 +153,124 @@ def netmask
...
+        frulist
+      rescue Exception => e
+        msg = "Unknown error getting fru list."
+        begin
+          # Surprisingly, this works around an undocumented feature or bug with
+          # freeipmi on IBM/Lenovo servers. The first time around, freeipmi
+          # returns "invalid byte sequence in UTF-8". The second time does a
+          # successful fru list.
+          return host.fru.list if host.is_a? Rubyipmi::Freeipmi::Connection
+        rescue
```
Comment (dmitri-d) [inline modules/bmc/ipmi.rb]: The call above can throw RuntimeErrors only?

File: modules/bmc/ipmi.rb [hunk 3]
```diff
@@ -148,6 +153,124 @@ def netmask
...
+      end
+
+      # print all FRU information
+      def frulist
+        frulist = host.fru.list
+        # Detect quirk on ipmitool
+        raise 'Quirk' if frulist.empty?
+        frulist
+      rescue Exception => e
+        msg = "Unknown error getting fru list."
```
Comment (dmitri-d) [inline modules/bmc/ipmi.rb]: Is this the case? Wouldn't exception contain the details of the error that caused it?

File: modules/bmc/ipmi.rb [hunk 4]
```diff
@@ -148,6 +153,124 @@ def netmask
...
+          # freeipmi on IBM/Lenovo servers. The first time around, freeipmi
+          # returns "invalid byte sequence in UTF-8". The second time does a
+          # successful fru list.
+          return host.fru.list if host.is_a? Rubyipmi::Freeipmi::Connection
+        rescue
+          return msg
+        end
+        # Handle ipmitool
+        return "#{msg} Try bmc_provider=freeipmi for possible quirk workaround." if e.message == 'Quirk'
+        return msg
```
Comment (dmitri-d) [inline modules/bmc/ipmi.rb]: This is also hard to follow. I'd suggest returning an error message at the top of the rescue block.

**Review by dmitri-d — CHANGES_REQUESTED** [review]

File: modules/bmc/ipmi.rb (hunk 1 above)
Comment (Deltik) [reply modules/bmc/ipmi.rb]: This has been fixed in 7fbe331.

File: modules/bmc/ipmi.rb (hunk 3 above)
Comment (Deltik) [reply modules/bmc/ipmi.rb]: This has been fixed in 7fbe331.  Exceptions are now returned as

> Error getting fru list: `Exception#message`

File: modules/bmc/ipmi.rb (hunk 2 above)
Comment (Deltik) [reply modules/bmc/ipmi.rb]: I'm catching all exceptions and feeding their messages through to the client.

File: modules/bmc/ipmi.rb (hunk 4 above)
Comment (Deltik) [reply modules/bmc/ipmi.rb]: This has been fixed in 7fbe331.

**Comment by dmitri-d** [comment]
[test]

**Review by dmitri-d — APPROVED** [review]

**Comment by dmitri-d** [comment]
Looks good, waiting for the results of the build.

**Comment by dmitri-d** [comment]
Merged, thanks @Deltik.

**Comment by dmitri-d** [comment]
@Deltik: did you have anything in mind re: changes to Foreman UI?

**Comment by Deltik** [comment]
@witlessbird: Foreman could use this additional functionality as part of
> Foreman » Hosts » All hosts » (any host with a BMC) » BMC

Foreman hosts BMC info

Currently, when showing any host with a BMC, Foreman makes these requests in the background to the associated Smart Proxy:

- `GET /bmc/BMC_IP_ADDRESS/chassis/power/status HTTP/1.1`
- `GET /bmc/BMC_IP_ADDRESS/lan/ip HTTP/1.1`
- `GET /bmc/BMC_IP_ADDRESS/lan/netmask HTTP/1.1`
- `GET /bmc/BMC_IP_ADDRESS/lan/mac HTTP/1.1`
- `GET /bmc/BMC_IP_ADDRESS/lan/gateway HTTP/1.1`

, where `BMC_IP_ADDRESS` is the IP address of the BMC.

I'm thinking there could be buttons added to fetch additional information from BMC that take longer than a few seconds, such as `GET /bmc/BMC_IP_ADDRESS/fru/list`, which takes 15 seconds for me using ipmitool and 2 seconds for me using freeipmi.

There could also be a button to fetch sensor information.  And a button to perform a BMC reset.

I haven't thought this through much, but Foreman can use these new API methods for additional BMC functionality.

---

We can also fix an issue with this:

Foreman host power actions

Currently, "Cycle" means the same thing as "Reset" because they both call `PUT /bmc/BMC_IP_ADDRESS/chassis/power/cycle`, but now, "Reset" can be changed to mean `/bmc/BMC_IP_ADDRESS/chassis/power/reset`.
